### PR TITLE
Payable payout

### PR DESCRIPTION
**Pull request by jiqiang90**

### CODE REVIEW

File: schema.graphql [hunk 1]
```diff
@@ -269,5 +270,5 @@ type PayoutDetail @entity {
   account: Account!
   amount: BigInt!
   isClaimed: Boolean!
-  payout: ValidatorPayout!
+  payout: ValidatorPayout #if it is not claimed, it has no validator payout
```
Comment (ERussel) [inline schema.graphql:273]: If this field matches whether reward was claimed or not then `isClaimed` looks redundant

File: src/mappings/stakingHandlers.ts [hunk 2]
```diff
@@ -3,38 +3,55 @@ import { sha256 } from 'js-sha256';
...
+        }else{
+            await saveCurrentEra(currentEra, currentBlockNumber);
+            await saveValidators(currentEra.toNumber());
+        }
+    }
 }
 
-async function saveCurrentEra(currentEra: EraIndex, blockNumber: number) {
+async function saveCurrentEra(currentEra: EraIndex, blockNumber: number){
     let currentEraNumber = currentEra.toNumber()
```
Comment (ERussel) [inline src/mappings/stakingHandlers.ts:45]: Maybe change to `currentEraIndex` because we have `previousEraIndex`

**Review by ERussel — APPROVED** [review]

**Review by ef1rspb — APPROVED** [review]

File: schema.graphql (hunk 1 above)
Comment (jiqiang90) [reply schema.graphql:273]: good point

File: src/mappings/stakingHandlers.ts [hunk 3]
```diff
@@ -3,38 +3,55 @@ import { sha256 } from 'js-sha256';
...
+import { uniq, flatten } from "lodash";
 
-export async function handleSession(event: SubstrateEvent): Promise<void> {
-    const currentEraOptional = await api.query.staking.currentEra();
-    if (currentEraOptional.isNone) return;
+async function ensureSession(sessionId: number, startBlock: number){
+    let session = new Session(sessionId.toString());
+    session.startBlock = startBlock;
+    await session.save()
+    let previousSession = await Era.get((sessionId-1).toString());
```
[the remark this answers is not in the document]
Comment (jiqiang90) [reply src/mappings/stakingHandlers.ts:17]: In that case previousSession return undefined, if will not update endblock for previous session entity. By the way, it should be `Session.get(...`

File: src/mappings/stakingHandlers.ts [hunk 4]
```diff
@@ -88,3 +105,178 @@ async function saveEraValidator(currentEraNumber: number, exposure: Exposure, va
...
+
+    if (!event.extrinsic){
+        logger.warn(`Reward event ${event.block.block.header.number}-${event.idx} has no extrinsic`)
+        await checkPayoutEraEnd(account.toString(),event.block.block.header.number.toString());
+    } else if(!event.extrinsic.success){
+        return
+    }
+    else{
+        const extrinsic = event.extrinsic;
+        const call = extrinsic.extrinsic.method;
```
[the remark this answers is not in the document]
Comment (jiqiang90) [reply src/mappings/stakingHandlers.ts:152]: `const call = extrinsic.extrinsic.method;` is correct

File: src/mappings/stakingHandlers.ts [hunk 5]
```diff
@@ -88,3 +105,178 @@ async function saveEraValidator(currentEraNumber: number, exposure: Exposure, va
...
+    } = event;
+    const era = await Era.get(eraIndex.toString());
+    if(era){
+        await syncEraPayOut(eraIndex as EraIndex,validatorPayout as BalanceOf)
+    }else{
+        logger.error(`EraPayout at era ${eraIndex.toString()},but can not find era`)
+        process.exit(1)
+    }
+}
+async function syncEraPayOut(eraIndex: EraIndex, eraTotalPayout: BalanceOf): Promise<void>{
```
[the remark this answers is not in the document]
Comment (jiqiang90) [reply src/mappings/stakingHandlers.ts:123]: ?

**Pull request merged**